bindings/python/test/test_numpy.py

```diff
@@ -91,3 +95,84 @@ def test_aggregate_simple(self):
         assert len(agg_cmd.command["pipeline"]) == 2
         self.assertEqual(agg_cmd.command["pipeline"][0]["$project"], projection)
         self.assertEqual(agg_cmd.command["pipeline"][1]["$project"], {"_id": True, "data": True})
+
+    def round_trip(self, data, schema, coll=None):
+        if coll is None:
+            coll = self.coll
+        coll.drop()
+        res = write(self.coll, data)
+        self.assertEqual(len(list(data.values())[0]), res.raw_result["insertedCount"])
+        self.assert_numpy_equal(find_numpy_all(coll, {}, schema=schema), data, schema=schema)
+        return res
+
+    def schemafied_ndarray_dict(self, dict, schema):
+        ret = {}
+        for k, v in dict.items():
+            ret[k] = np.array(v, dtype=schema[k])
+        return ret
+
+    def test_write_error(self):
+        schema = {"_id": "int32", "data": "int64"}
+        length = 10001
+        data = {"_id": [i for i in range(length)] * 2, "data": [i * 2 for i in range(length)] * 2}
+        data = self.schemafied_ndarray_dict(data, schema)
+        with self.assertRaises(ArrowWriteError):
+            try:
+                self.round_trip(data, Schema({"_id": int32(), "data": int64()}))
+            except ArrowWriteError as awe:
+                self.assertEqual(
+                    10001, awe.details["writeErrors"][0]["index"], awe.details["nInserted"]
+                )
+                raise awe
+
+    def test_write_schema_validation(self):
```

Comment:
I can't help but feel like all this "schema" management is painful. Should a schema even be required for reading? It should be fairly straightforward to auto-discover the schema based on the fields and types in the first document. Not really an issue in this PR itself but something to think about. Do we already have a ticket for this auto-discover schema idea?